### ISSUE TITLE
Should we keep the 12.5 syntax as is?

### ISSUE THREAD

**Issue by coderanger** [chef/chef#4672]
Opening this here, though maybe this would need to be an RFC. The 12.5 action syntax is cool, easy to teach, really convenient, and full of pitfalls you could drive a bus through. I think we should consider it an evolutionary dead end and try to come up with something more workable over the long term. The root of the problem is that inside the scope for actions we are cramming the provider methods, the resource methods (including resource properties), and the full recipe DSL. I think we need to find a way to disentangle these, but we should probably discuss this a bunch before doing anything. What say you all?

**Comment by lamont-granquist** [comment]
:+1:, but now we already have legacy code to support and need to disentangle it carefully.

we probably need a slightly more general concept than use_inline_resources which is a declaration of which API version your resource is compatible with.

we may want to consider having a cleanroom resource compiler class which only has the methods on the resource dsl in it.  then we don't wind up automatically injecting the binding of the resource we're constructing globally into the whole dsl.  this would have other useful features since we could spool up the methods that are called on the dsl and then create the two classes and play the methods back on them, and that would allow us to setup resource and provider inheritance via methods on the dsl.

i was wanting this before, but thought it was a lot of work just to get inheritance.  now i kinda think it might be useful to clean up the awful binding issues...

**Comment by stevendanna** [comment]
> full of pitfalls you could drive a bus through

Do we have a lit of these somewhere?

**Comment by jkeiser** [comment]
Yeah. The nested problem is way gnarlier than it looked. I think we need to work back from the list of things it would be nice to accomplish ("ability to copy/paste port an existing provider" being one of them), in priority, and attack it from that direction. While I don't think we can now maintain perfect backcompat, I bet we can get darn close, especially since most of the ambiguities now come with warnings--"backcompat with any code that does not currently trigger a warning" is not an unreasonable rubrick.

The biggest pitfall of the present design I'm aware of is that using naked `property_name` changes its meaning when you are creating a resource inside an action--*sometimes*. And it happens in some super compelling scenarios, like when you're wrapping a resource with similar properties:

```ruby
resource_name :user_config_file
property :owner, String
property :path_to_widget, String
action :create do
  # owner in the next line is fine, because it's not inside the file
  file "/home/#{owner}/.config" do
    # oops: owner doesn't mean the same thing anymore, even though it's sort of on the same resource
    owner owner
    # this is perfectly fine, and does what you expect, even though owner doesn't.
    content "path_to_widget #{path_to_widget}"what you expect!
  end
end
```

**Comment by jkeiser** [comment]
Right now, BTW, we generate warnings if you do any of these. We could actually upgrade them to errors if you are using a custom resource, which would mitigate the crap out of the custom resource problem, and give people an actionable way to fix it. (I am aware that the warnings are appearing in places they shouldn't right now, it's on the list :)

**Comment by coderanger** [comment]
To fully illustrate the other major one:

```ruby
property :user

action :run do
  user user
end
```

**Comment by coderanger** [comment]
@thommay If you're running the meeting tomorrow, can we put this on the agenda?

**Comment by jkeiser** [comment]
The spurious warnings are gone (or will be) in 12.9.

**Comment by lamont-granquist** [comment]
I'm not quite certain what the root cause of the tomcat issue was above, but it highlights a fundamental problem where we've got both the properties 'API' in the resource and the recipe DSL injected into the binding -- the result was that the tomcat cookbook failed only when the firewall cookbook was also in the run_list.  Having the recipe DSL in between what you are typing in the action and the resources makes for lots of cookbook cross-talk excitement.

**Comment by lamont-granquist** [comment]
That also highlights why I think I really hate the warnings.

As a cookbook author you'd need to test your cookbook against every other cookbook that has been published in order to make sure it works with all possible combinations.  As a user, the UX is that you'll use a cookbook with another cookbook and wind up with a pile of warnings and possibly an exception and it'll fail and you have to track down why.
